internal/controller/ipam_controller.go

```diff
@@ -0,0 +1,204 @@
+/*
+ Copyright 2025, NVIDIA CORPORATION & AFFILIATES
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/Mellanox/spectrum-x-operator/pkg/config"
+
+	nvipamv1 "github.com/Mellanox/nvidia-k8s-ipam/api/v1alpha1"
+)
+
+const (
+	spectrumXIPAMControllerName = "spectrumxipamcontroller"
+	cidrPoolType                = "CIDRPool"
+	spectrumXFinalizer          = "spectrumx.nvidia.com/ipam"
+)
+
+// NvIPAMReconciler reconciles Spectrum-X-Config from ConfigMap
+type NvIPAMReconciler struct {
+	client.Client
+	Scheme             *runtime.Scheme
+	ConfigMapNamespace string
+	ConfigMapName      string
+	CIDRPoolsNamespace string
+}
+
+//+kubebuilder:rbac:groups="",resources=configmaps/finalizers,verbs=update
+//+kubebuilder:rbac:groups=nv-ipam.nvidia.com,resources=cidrpools,verbs=get;list;patch;delete;deletecollection
+
+func (r *NvIPAMReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logr := log.FromContext(ctx)
+	logr.Info("Reconciling Spectrum-X ConfigMap")
+	// Fetch the ConfigMap
+	cm := &corev1.ConfigMap{}
+	err := r.Get(ctx, req.NamespacedName, cm)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			// Return early if the object is not found.
+			return ctrl.Result{}, nil
+		}
+	}
+	// Handle deletion reconciliation loop.
+	if !cm.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cm)
+	}
+
+	spectrumXConfig, err := config.ParseConfig(cm.Data[config.ConfigMapKey])
+	if err != nil {
+		logr.Info("Failed to parse Spectrum-X ConfigMap Data")
+		return ctrl.Result{}, err
+	}
+	logr.Info("Parsed Spectrum-X Config", "config", spectrumXConfig)
+
+	err = r.reconcileCIDRPools(ctx, spectrumXConfig)
+	if err != nil {
+		logr.Info("Failed to reconcile CIDRPool")
+		return ctrl.Result{}, err
+	}
+	// Add finalizer if not set.
+	if !controllerutil.ContainsFinalizer(cm, spectrumXFinalizer) {
```

Comment:
should we handle finalizers before reconciling CIRDPools?

Reply:
Done